### PR TITLE
Update to go-ethereum 1.8.1

### LINKED ISSUE #638: Switch code from whisperV5 to whisperV6
# Problem

Switch code from whisperV5 to whisperV6

# Implementation

Change imports from "github.com/ethereum/go-ethereum/whisper/whisperv5" to `whisper "github.com/ethereum/go-ethereum/whisper/whisperv6` in status-go code.
Also make a patch to change imports in Geth packages: `go-ethereum\whisper\notifications` and `go-ethereum\whisper\mailserver`.

# Acceptance Criteria

All tests are passed. Patch for go-ethereum packages added.

Part of #634

### LINKED ISSUE #665: Investigate/perform upgrade to Geth 1.8.1
# Problem

Geth 1.8.1 was [released today](https://blog.ethereum.org/2018/02/14/geth-1-8-iceberg%C2%B9/). It contains many improvements that are important for us, and by upgrading to 1.8.1 we render some patches redundant (namely `0011-geth-17-whisperv6-70fbc87.patch` which ported Whisper v6 to our fork), making future maintenance easier.

# Implementation

Investigate breaking changes in 1.8.1 to determine what work must be done as part of the upgrade. Perform upgrade.

## Breaking changes checklist
- [x] Discovery v4 and v5 have been merged to use the same UDP port (30303 by default). If you are doing manual peer management and using the light client, you may need to ensure your v1.8.0 clients are pointed to port 30303 and not 30304 as previously.
- [x] Trie pruning is enabled on all `--syncmode` variations (including `--syncmode=full`). If you are running an archive node where you would like to retain all historical data, you should disable pruning via `--gcmode=archive`.
- [x] Only the latest 128 tries are kept in memory, most tries are garbage collected. If you are running a block explorer or other service relying on transaction tracing without an archive node (`--gcmode=archive`), you need to trace within this window! Alternatively, specify the `reexec: 12345` tracer option to allow regenerating historical state; and ideally switch to chain tracing which amortizes overhead across all traced blocks.
- [x] Native events rely on modifications to internal go-ethereum types within generated code. If you are using wrappers generated prior to v1.8.0, you will need to regenerate them to be compatible with the new code base.
- [x] The HTTP/WS RPC endpoint was extended with [DNS rebind protection](https://github.com/ethereum/go-ethereum/pull/15962). If you are running an RPC endpoint **addressed by name rather than IP**, run with `--rpcvhost=your.domain` to continue accepting remote requests.

# Acceptance Criteria

- [ ] The different `make` targets work as expected (e.g. `make ci`, `make statusgo-android`, etc)
- TBD

# Notes

This GHI only makes sense to start once https://github.com/status-im/status-go/pull/660 is merged.

TDB

# Future Steps

- Upgrade bootnodes and mailserver cluster with new version.
- TBD

### PR DESCRIPTION
**Pull request by pedropombeiro**
This PR updates our vendored geth to 1.8.1, while making sure that any updates to whisper v6 in upstream are reverted to v5.

This involved:
- [x] Deleting `0011-geth-17-whisperv6-70fbc87.patch`, since v6 no longer needs patching
- [x] Remove DNS rebinding protection for localhost, so that e2e tests pass
- [x] Address type and method signature changes 
- [x] Add patch to downgrade usage of Whisper v6 to v5 in some geth 1.8.1 vendor files
- [x] Add patches required for cross-compiled builds starting with geth 1.8.0.

Changes of interest in 1.8.1:
- Geth now writes less state to disk, slowing down datadir growth ([#15857](https://github.com/ethereum/go-ethereum/pull/15857))
- The RPC server now protects against DNS rebinding attacks ([#15962](https://github.com/ethereum/go-ethereum/pull/15962))
- The experimental peer discovery v5 protocol has changed, breaking backwards compatibility with previous releases. Geth 1.8 will not discover LES servers running geth 1.7.
- Several bugs around CHT handling are resolved (#15692, #16039)
- Light client peer count is now properly limited ([#16010](https://github.com/ethereum/go-ethereum/pull/16010))
- Whisper protocol v6 as described in EIP-627 is implemented (#15666, #15701, #15802, #15578, #15631, #16046)
- Discovery v4 and the experimental discovery v5 protocol now run on the same UDP port ([#15200](https://github.com/ethereum/go-ethereum/pull/15200))
- Several issues in the experimental discovery v5 protocol are resolved (#15995, #16036)
- The `GasLimit` and `GasUsed` fields of blocks and transactions have type `uint64` instead of `*big.Int`. **This is a breaking API change**. ([#15466](https://github.com/ethereum/go-ethereum/pull/15466))

Closes #665

### CODE REVIEW

**Comment by mandrigin** [comment]
@PombeirP what about other `geth-17*` patches? They need to be removed as well...

**Comment by pedropombeiro** [comment]
@mandrigin you're right, I was so focused on making the transition to 1.8.1 work well that in the end I overlooked removing the 1.7 patches. But I'm not sure that the one 1.7-only patch can be removed. If you see the differences, the behavior upstream is not the same as in our patch:

image

So maybe I need to rename this patch to `0010-geth-18-fix-npe-in-filter-system.patch`?

**Comment by mandrigin** [comment]
@PombeirP hmm that's weird, I'm a pretty much sure I saw that fixed on master... I'll double check.

File: geth/common/utils.go [hunk 1]
```diff
@@ -143,6 +144,11 @@ func Fatalf(reason interface{}, args ...interface{}) {
 
 // CreateTransaction returns a transaction object.
 func CreateTransaction(ctx context.Context, args SendTxArgs) *QueuedTx {
+	if args.Input == nil && args.Data == nil {
+		// If neither Input nor Data were initialized, let's pass an empty slice in Input
+		emptyByteArray := hexutil.Bytes(make([]byte, 0))
+		args.Input = &emptyByteArray
```
Comment (adambabik) [inline geth/common/utils.go]: If `Input` is a pointer why can't it be `nil`?

File: geth/common/rpccall.go [hunk 2]
```diff
@@ -105,12 +105,13 @@ func (r RPCCall) ParseGas() *hexutil.Big {
...
 	}
 
-	parsedValue, err := hexutil.DecodeBig(inputValue)
+	parsedValue, err := hexutil.DecodeUint64(inputValue)
 	if err != nil {
 		return nil
 	}
 
-	return (*hexutil.Big)(parsedValue)
+	_v := hexutil.Uint64(parsedValue)
```
Comment (adambabik) [inline geth/common/rpccall.go]: `_v` is a bit strange in Go. `v := hexutil.Uint64(parsedValue)` looks ordinary.

File: geth/common/types.go [hunk 3]
```diff
@@ -163,11 +163,14 @@ type QueuedTx struct {
...
 	To       *common.Address `json:"to"`
-	Gas      *hexutil.Big    `json:"gas"`
+	Gas      *hexutil.Uint64 `json:"gas"`
 	GasPrice *hexutil.Big    `json:"gasPrice"`
 	Value    *hexutil.Big    `json:"value"`
-	Data     hexutil.Bytes   `json:"data"`
 	Nonce    *hexutil.Uint64 `json:"nonce"`
+	// We accept "data" and "input" for backwards-compatibility reasons. "input" is the
+	// newer name and should be preferred by clients.
+	Data  *hexutil.Bytes `json:"data"`
```
Comment (adambabik) [inline geth/common/types.go]: Do we need to change the data type to `*hexutil.Bytes`?

File: geth/common/types.go [hunk 4]
```diff
@@ -163,11 +163,14 @@ type QueuedTx struct {
 type SendTxArgs struct {
 	From     common.Address  `json:"from"`
 	To       *common.Address `json:"to"`
-	Gas      *hexutil.Big    `json:"gas"`
+	Gas      *hexutil.Uint64 `json:"gas"`
 	GasPrice *hexutil.Big    `json:"gasPrice"`
 	Value    *hexutil.Big    `json:"value"`
-	Data     hexutil.Bytes   `json:"data"`
 	Nonce    *hexutil.Uint64 `json:"nonce"`
+	// We accept "data" and "input" for backwards-compatibility reasons. "input" is the
```
Comment (adambabik) [inline geth/common/types.go]: It's our internal struct, so maybe we can remove all `Data` field and fix all occurrences in this PR?

File: geth/common/utils.go [hunk 5]
```diff
@@ -143,6 +144,11 @@ func Fatalf(reason interface{}, args ...interface{}) {
 
 // CreateTransaction returns a transaction object.
 func CreateTransaction(ctx context.Context, args SendTxArgs) *QueuedTx {
+	if args.Input == nil && args.Data == nil {
```
Comment (adambabik) [inline geth/common/utils.go]: I don't get why we need this condition.

File: geth/transactions/txqueue_manager.go [hunk 6]
```diff
@@ -207,14 +208,21 @@ func (m *Manager) completeTransaction(config *params.NodeConfig, selectedAccount
 	}
 
 	chainID := big.NewInt(int64(config.NetworkID))
-	data := []byte(args.Data)
+	input := args.Input
+	if input == nil {
+		input = args.Data
+	}
+	if input == nil {
+		return hash, errors.New("Missing Input/Data byte array")
```
Comment (adambabik) [inline geth/transactions/txqueue_manager.go]: That should be totally fine. Do you get an error if `data` passed to `tx := types.NewTransaction(nonce, toAddr, value, gas, gasPrice, data)` is `nil`?

**Review by adambabik — CHANGES_REQUESTED** [review]

**Comment by adambabik** [comment]
@PombeirP Overall it looks good. I am not really sure about these changes to `SendTxArgs`. It's our struct and we have full control over it so we can clean it immediately without having two fields to store data/input. Also, it looks like it may be beneficial to change its `Gas` field type to non-pointer.

File: geth/common/rpccall.go (hunk 2 above)
Comment (pedropombeiro) [reply geth/common/rpccall.go]: Will fix

File: geth/common/utils.go (hunk 1 above)
Comment (pedropombeiro) [reply geth/common/utils.go]: @adambabik the changes to the `SendTxArgs` struct were based on the fact that is looked like a clone of the one in `vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go`, so I assumed we wanted to keep them in sync (even the comments in our struct are identical). If we have the freedom over our own struct, we should have a comment on it that clarifies that. I'll add one for your review.

File: geth/common/types.go (hunk 3 above)
Comment (pedropombeiro) [reply geth/common/types.go]: I don't think so, I was trying to keep it similar to geth's struct. Will change it to `Input hexutil.Bytes`.

File: geth/common/utils.go (hunk 5 above)
Comment (pedropombeiro) [reply geth/common/utils.go]: This was more of a sanity check during the migration that slipped through.

**Comment by pedropombeiro** [comment]
@adambabik thanks for the review so far. Regarding `Gas`, what you're suggesting is to use `0` in cases where `nil` was being used (e.g. `RPCCall.ParseGas`)?

**Comment by pedropombeiro** [comment]
Adding missing tests on bloom filter logic.

**Review by adambabik — APPROVED** [review]
Much cleaner 👏  LGTM

File: geth/common/rpccall.go [hunk 7]
```diff
@@ -94,23 +94,23 @@ func (r RPCCall) ParseValue() *hexutil.Big {
 
 // ParseGas returns the hex big associated with the call.
 // nolint: dupl
-func (r RPCCall) ParseGas() *hexutil.Big {
+func (r RPCCall) ParseGas() hexutil.Uint64 {
```
Comment (mandrigin) [inline geth/common/rpccall.go]: Does it make any sense to actually return also an error here, what do you think?

File: geth/transactions/ethtxclient.go [hunk 8]
```diff
@@ -84,8 +84,8 @@ func toCallArg(msg ethereum.CallMsg) interface{} {
 	if msg.Value != nil {
 		arg["value"] = (*hexutil.Big)(msg.Value)
 	}
-	if msg.Gas != nil {
-		arg["gas"] = (*hexutil.Big)(msg.Gas)
+	if msg.Gas != 0 {
```
Comment (mandrigin) [inline geth/transactions/ethtxclient.go:87]: Is `0` for sure an invalid value for Gas?
Before the logic assumed that there are 3 types of values: non-set (`nil`), set, but 0, set, != 0.
Here we lost one state.

Comment (mandrigin) [reply geth/transactions/ethtxclient.go:87]: if so I'd extracted it to a constant to simplify changes in the future

File: geth/transactions/txqueue_manager.go [hunk 9]
```diff
@@ -207,31 +207,32 @@ func (m *Manager) completeTransaction(config *params.NodeConfig, selectedAccount
...
 			return hash, err
 		}
-		if gas.Cmp(big.NewInt(defaultGas)) == -1 {
+		if gas < defaultGas {
 			log.Info("default gas will be used. estimated gas", gas, "is lower than", defaultGas)
-			gas = big.NewInt(defaultGas)
+			gas = defaultGas
 		}
+	} else {
+		gas = uint64(args.Gas)
```
Comment (mandrigin) [inline geth/transactions/txqueue_manager.go]: what if `gas` set here is < `defaultGas`?

File: _assets/patches/geth/README.md [hunk 10]
```diff
@@ -34,9 +34,10 @@ Instructions for creating a patch from the command line:
 - [`0006-latest-cht.patch`](./0006-latest-cht.patch) – updates CHT root hashes, should be updated regularly to keep sync fast, until proper Trusted Checkpoint sync is not implemented as part of LES/2 protocol.
 - [`0009-whisper-envelopes-tracing.patch`](./0009-whisper-envelopes-tracing.patch) — adds Whisper envelope tracing (need to be reviewed and documented)
 - [`0010-geth-17-fix-npe-in-filter-system.patch`](./0010-geth-17-fix-npe-in-filter-system.patch) - Temp patch for 1.7.x to fix a NPE in the filter system.
-- [`0011-geth-17-whisperv6-70fbc87.patch`](./0011-geth-17-whisperv6-70fbc87.patch) - Temp patch for 1.7.x to update whisper v6 to the upstream version at the `70fbc87` SHA1.
 - [`0014-whisperv6-notifications.patch`](./0014-whisperv6-notifications.patch) — adds Whisper v6 notifications (need to be reviewed and documented)
 - [`0015-whisperv6-envelopes-tracing.patch`](./0015-whisperv6-envelopes-tracing.patch) — adds Whisper v6 envelope tracing (need to be reviewed and documented)
+- [`0017-geth-18-downgrade-to-whisperv5.patch`](./0017-geth-18-downgrade-to-whisperv5.patch) — some files in geth 1.8 import Whisper v6, instead of v5. This patch ensures v5 is used everywhere
```
Comment (mandrigin) [inline _assets/patches/geth/README.md:39]: Can we also have a word here about the xgo patches?

**Review by mandrigin — CHANGES_REQUESTED** [review]
👍 huge job! just a few questions...

File: geth/transactions/ethtxclient.go (hunk 8 above)
Comment (pedropombeiro) [reply geth/transactions/ethtxclient.go:87]: Hence my question above (https://github.com/status-im/status-go/pull/702#issuecomment-368442046). I'm not against keeping the old version that uses `nil`, tbh.

File: _assets/patches/geth/README.md (hunk 10 above)
Comment (pedropombeiro) [reply _assets/patches/geth/README.md:39]: Did you see the README.md file in the `geth-xgo` folder? Or are you just saying we should add a mention here to that folder?

Comment (mandrigin) [reply _assets/patches/geth/README.md:39]: ah, missed that file. it's fine then.

File: geth/transactions/ethtxclient.go (hunk 8 above)
Comment (pedropombeiro) [reply geth/transactions/ethtxclient.go:87]: Reverted the changes to `Gas` type. I don't think that we gain much by doing that.

**Review by mandrigin — APPROVED** [review]
⭐️

**Comment by PombeirP** [comment]
Thanks @adambabik. Can you mark the review as approved then?

**Pull request merged**